fenix/app/src/main/java/org/mozilla/fenix/compose/ComposableFactory.kt

```diff
@@ -0,0 +1,75 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.compose
+
+import androidx.compose.foundation.background
+import androidx.compose.foundation.layout.Box
+import androidx.compose.foundation.layout.size
+import androidx.compose.runtime.Composable
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.layout.ContentScale
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.unit.dp
+import mozilla.components.concept.base.images.ImageLoadRequest
+import org.mozilla.fenix.components.components
+import org.mozilla.fenix.theme.FirefoxTheme
+import org.mozilla.fenix.compose.ThumbnailImage as ThumbnailImageContent
+
+/**
+ * Factory object used to instantiate Composables with preview-breaking dependencies, such as
+ * [components].
+ */
+object ComposableFactory {
+
+    /**
+     * Loads a thumbnail belonging to a [ImageLoadRequest], fetching asynchronously the bitmap from storage.
+     * If this is in a preview, an empty box will be rendered instead.
+     *
+     * @param request [ImageLoadRequest] used to fetch the thumbnail bitmap.
+     * @param contentScale [ContentScale] used to draw image content.
+     * @param alignment [Alignment] used to draw the image content.
+     * @param modifier [Modifier] used to draw the image content.
+     * @param fallbackContent The content to display with a thumbnail is unable to be loaded.
+     */
+    @Composable
+    fun ThumbnailImage(
```

Comment:
Noah and I were brainstorming about this yesterday. Here's some of the reasoning for something along this route:

1. Modularization. Modules won't have access to `components`, so they won't be able to have implicit, hidden dependencies on it like is in the `ThumbnailImage` child Composable below. This can result in "prop drilling", where we pass these dependencies down the entire Composable stack. 
2. Even before modularization, the `components` dependency is quite hidden in the stack here. I am not strictly against Composables that can retrieve their own dependencies, but they should then fully own those dependencies.

Using a factory pattern seemed like it might help reveal the hidden dependency a bit by requiring a step "outside" of the usual Compose world, but I don't think this is the only solution to these problems. In fact, there is probably quite a bit of prior art from the React and Redux communities that could inform our solution. @boek may have more insight there.

Ultimately, I would like to see us come to a team agreement on how to solve this problem as I think it would help the team iterate faster and avoid disparate novel solutions. That said, I am in no way married to anything that as presented in this patch.

Reply:
Jeff and I spoke offline about this, and I have additional suggestions to share when I get some time tomorrow

Reply:
Apologies for coming back a little late to this.

My short-term suggestion to avoid the preview breakage would be to pass the dependency explicitly.

Longer-term, the strategy I think we should investigate but would require team buy-in

## FakeComponents

1. Create an abstract supertype of `Components`
2. Create fakes for each component type 
3. update the globally accessible `@Composable components` to be something like: `get() = if (inComposePreview) LocalContext.current.components else FakeComponents()`

I also think we should consider moving the `components` from the `LocalContext` to the `CompositionLocal` to distance their association from the application context. It would also give us more control over local overrides, which is risky but potentially useful.

Then, I think it is fine for some Composables to retrieve dependencies **for their children** from the local composition. This would be required for Composables defined in modules that won't have knowledge of `Components`, and also keeps dependencies more explicit as well as practicing a bit of inversion of control. We would still have implicit dependencies hidden in the tree in the Composables that retrieved components for their children, but I think that is an okay trade-off as long as we limit it to our normally globally-accessible types for now.

A fair amount of stuff is touched on in the [docs for CompositionLocal](https://developer.android.com/jetpack/compose/compositionlocal). For example, we could use the fake types as sane defaults. 

As an aside: an interesting case is that they suggest _not_ using `CompositionLocal` to hold `ViewModel`s, which makes sense as they are screen-specific. However, if we move to a full Redux model with a single store, I think that we could get away with injecting that through `CompositionLocal`, which means we could make more interesting decisions about what level of the tree we observed Store properties and define action dispatches at.

Reply:
Ah, Noah reminded me that we could just do the `inComposePreview` check in `TabThumbnail` directly and access `components` there. I think that's fine for now, but that we'll want a better long-term solution that is well-defined and can be re-used across the team.

> Tangential on modularization in case we haven't thought of this already, compose/designsystem package should not contain any feature specific components like TabThumbnail to not have dependency on feature models.

`TabThumbnail` in particular looks to only have dependencies on lower-level types like `ThumbnailStorage` and `TabSessionState`. Given our domain, I imagine we could potentially include things like a `TabThumbnail` in our design system, or at the very least in a module between the design system and the app module. I admit it's been a while since I've looked over our design system tokens, so are we imagining a module that contains only components that are agnostic of the browser domain?